flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -227,6 +230,25 @@ arrayElementToExtendedValue(Fortran::lower::FirOpBuilder &builder,
       [&](const auto &) -> fir::ExtendedValue { return element; });
 }
 
+/// Build the ExtendedValue for \p element that is an element of an array or
+/// array section with \p array base (`array` or `array(i:j:k)%x%y`).
+/// If it is an array section, \p slice must be provided and be a fir::SliceOp that
+/// describes the section.
+static fir::ExtendedValue
+arraySectionElementToExtendedValue(Fortran::lower::FirOpBuilder &builder,
+                            mlir::Location loc, const fir::ExtendedValue &array,
+                            mlir::Value element, mlir::Value slice) {
+  if (!slice)
+    return arrayElementToExtendedValue(builder, loc, array, element);
+  auto sliceOp = mlir::dyn_cast_or_null<fir::SliceOp>(slice.getDefiningOp());
+  assert(slice && "slice must be a sliceOp");
```

Comment:
I think the intent is to check the value of `sliceOp`, rather than `slice`, here.